From eaa161cf7d05e76d85d40acda7a3ff8ddf375623 Mon Sep 17 00:00:00 2001
From: Xuan <xgong@apache.org>
Date: Thu, 29 Oct 2015 17:36:36 -0700
Subject: [PATCH 1384/2748] YARN-4313. Race condition in MiniMRYarnCluster
 when getting history server address. Contributed
 by Jian He

(cherry picked from commit 7412ff48eeb967c972c19c1370c77a41c5b3b81f)
(cherry picked from commit b505bf9e582f5bca59be28296a4b4a640b3389eb)
(cherry picked from commit ceefabc44a76821f2efc1cc971969bc8952bc537)

Change-Id: Idae672060eeb4d9dab82a3fae0569a0f67f13756
---
 .../hadoop/mapreduce/v2/MiniMRYarnCluster.java     |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-jobclient/src/test/java/org/apache/hadoop/mapreduce/v2/MiniMRYarnCluster.java b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-jobclient/src/test/java/org/apache/hadoop/mapreduce/v2/MiniMRYarnCluster.java
index 47b38a1..207d0f6 100644
--- a/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-jobclient/src/test/java/org/apache/hadoop/mapreduce/v2/MiniMRYarnCluster.java
+++ b/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-jobclient/src/test/java/org/apache/hadoop/mapreduce/v2/MiniMRYarnCluster.java
@@ -190,6 +190,7 @@ public void serviceInit(Configuration conf) throws Exception {
     public JobHistoryServerWrapper() {
       super(JobHistoryServerWrapper.class.getName());
     }
+    private volatile boolean jhsStarted = false;
 
     @Override
     public synchronized void serviceStart() throws Exception {
@@ -211,9 +212,11 @@ public synchronized void serviceStart() throws Exception {
         new Thread() {
           public void run() {
             historyServer.start();
+            jhsStarted = true;
           };
         }.start();
-        while (historyServer.getServiceState() == STATE.INITED) {
+
+        while (!jhsStarted) {
           LOG.info("Waiting for HistoryServer to start...");
           Thread.sleep(1500);
         }
-- 
1.7.9.5

